src/main/java/net/dv8tion/jda/core/requests/Response.java

```diff
@@ -16,81 +16,112 @@
 
 package net.dv8tion.jda.core.requests;
 
+import java.io.BufferedReader;
+import java.io.Reader;
 import org.json.JSONArray;
-import org.json.JSONException;
 import org.json.JSONObject;
+import org.json.JSONTokener;
 
 public class Response
 {
     public static final int ERROR_CODE = -1;
-    public final Exception exception;
+    public static final String ERROR_MESSAGE = "ERROR";
+
     public final int code;
+    public final Exception exception;
+    public final String message;
     public final long retryAfter;
-    public final String responseText;
+    private final Object object;
+
+    protected Response(final Exception exception)
+    {
+        this.code = Response.ERROR_CODE;
+        this.message = Response.ERROR_MESSAGE;
+        this.object = null;
+        this.exception = exception;
+        this.retryAfter = -1;
+    }
 
-    protected Response(int code, String response, long retryAfter)
+    protected Response(final int code, final String message, final okhttp3.Response response, final long retryAfter)
     {
         this.code = code;
-        this.responseText = response;
+        this.message = message;
         this.exception = null;
         this.retryAfter = retryAfter;
+
+        if (response == null || response.body().contentLength() == 0)
+        {
+            this.object = null;
+            return;
+        }
+
+        try (Reader reader = response.body().charStream().markSupported() ? response.body().charStream()
+                : new BufferedReader(response.body().charStream())) // this doesn't add overhead as org.json would do that itself otherwise
+        {
+            char begin; // not sure if I really like this... but we somehow have to get if this is an object or an array
+            int mark = 1;
+            do
+            {
+                reader.mark(mark++);
+                begin = (char) reader.read();
+            }
+            while (Character.isWhitespace(begin));
+
+            reader.reset();
+
+            if (begin == '{')
+                this.object = new JSONObject(new JSONTokener(reader));
+            else if (begin == '[')
+                this.object = new JSONArray(new JSONTokener(reader));
+            else
+                this.object = null;
+        }
+        catch (final Exception e)
+        {
+            throw new RuntimeException("An error occured while parsing a RestAction the response", e);
```

Comment:
What happens if the response is neither JSONObject or JSONArray, like an xml response at some outage?